strapi/strapi
# Update AWS (aws-sdk) from V2 to V3 with TS

### What does it do?

It updates AWs S3 provider to v3. To avoid a breaking change I've added a method to extract the credentials from legacyOptions and s3Options, plus another method to mirror the AWS documentation, passing a credentials object to s3Options.
```
	// s3Object
  upload: {
    config: {
      sizeLimit: 250 * 1024 * 1024 * 1024, // 256mb in bytes
      provider: 'aws-s3',
      providerOptions: {
        s3Options: {
          accessKeyId: env('AWS_ACCESS_KEY_ID'),
          secretAccessKey: env('AWS_ACCESS_SECRET'),
          region: env('AWS_REGION'),
          params: {
            ACL: env('AWS_ACL', 'public-read'),
            signedUrlExpires: env('AWS_SIGNED_URL_EXPIRES', 15 * 60),
            Bucket: env('AWS_BUCKET'),
          },
        },
      },
      actionOptions: {
        upload: {},
        uploadStream: {},
        delete: {},
      },
    },
  },
```

```
	// s3Object with Credentials
  upload: {
    config: {
      sizeLimit: 250 * 1024 * 1024 * 1024, // 256mb in bytes
      provider: 'aws-s3',
      providerOptions: {
        s3Options: {
		 credentials: {
            accessKeyId: env('AWS_ACCESS_KEY_ID'),
            secretAccessKey: env('AWS_ACCESS_SECRET'),
		  },
          region: env('AWS_REGION'),
          params: {
            ACL: env('AWS_ACL', 'public-read'),
            signedUrlExpires: env('AWS_SIGNED_URL_EXPIRES', 15 * 60),
            Bucket: env('AWS_BUCKET'),
          },
        },
      },
      actionOptions: {
        upload: {},
        uploadStream: {},
        delete: {},
      },
    },
  },
```

```
	// Without s3Object
  upload: {
    config: {
      sizeLimit: 250 * 1024 * 1024 * 1024, // 256mb in bytes
      provider: 'aws-s3',

      providerOptions: {
          accessKeyId: env('AWS_ACCESS_KEY_ID'),
          secretAccessKey: env('AWS_ACCESS_SECRET'),
          region: env('AWS_REGION'),
          params: {
            ACL: env('AWS_ACL', 'public-read'),
            signedUrlExpires: env('AWS_SIGNED_URL_EXPIRES', 15 * 60),
            Bucket: env('AWS_BUCKET'),
          },
      },
      actionOptions: {
        upload: {},
        uploadStream: {},
        delete: {},
      },
    },
  },
```

### Why is it needed?

To fix the memory leak that the Data Transfer package has.

### How to test it?

1. Connect a provider following https://market.strapi.io/providers/@strapi-provider-upload-aws-s3
2. Test a remote transfer with Data Transfer to a remote destination that has AWS connected
3. Test that the Media Library still works as expected

### Related issue(s)/PR(s)
